[SPARK-27403][SQL] Fix `updateTableStats` to update table stats always with new stats or None

sujith71955 opened the pull request:
## What changes were proposed in this pull request?

System shall update the table stats automatically if user set spark.sql.statistics.size.autoUpdate.enabled as true, currently this property is not having any significance even if it is enabled or disabled. This feature is similar to Hives auto-gather feature where statistics are automatically computed by default if this feature is enabled.
Reference:
https://cwiki.apache.org/confluence/display/Hive/StatsDev

As part of fix , autoSizeUpdateEnabled  validation is been done initially so that system will calculate the table size for the user automatically and record it in metastore as per user expectation.

## How was this patch tested?
UT is written and manually verified in cluster.
Tested with unit tests + some internal tests on real cluster.

Before fix:

![image](https://user-images.githubusercontent.com/12999161/55688682-cd8d4780-5998-11e9-85da-e1a4e34419f6.png)



After fix
![image](https://user-images.githubusercontent.com/12999161/55688654-7d15ea00-5998-11e9-973f-1f4cee27018f.png)

sujith71955 committed:
[SPARK-274034][SQL] Table Statisics shall be updated automatically if auto update feature is enabled(spark.sql.statistics.size.autoUpdate.enabled =true)

What changes were proposed in this pull request?

For the  table, INSERT OVERWRITE command statistics are automatically computed by default if user set spark.sql.statistics.size.autoUpdate.enabled =true and the statistics shall be recorded
in metadata store, this is not happening currently  because of  validation table.stats.nonEmpty, the statistics were never recorded for the newly created table, this check  doesn't holds good if auto update property feature is enabled by the user.
As part of fix the autoSizeUpdateEnabled has been pulled up as part of separate validation which will ensure if this feature is enabled the system will calculate the size of the table in every insert command
and the same will be recorded in meta-store.

How was this patch tested?
UT is written and manually verified in cluster.

Tested with unit tests + some internal tests on real cluster.

SparkQA commented:
**[Test build #104362 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/104362/testReport)** for PR 24315 at commit [`b143e84`](https://github.com/apache/spark/commit/b143e845f9281c44d69900282b32e6ac99eb2cb0).

sujith71955 commented on sql/core/src/main/scala/org/apache/spark/sql/execution/command/CommandUtils.scala:
```diff
@@ -42,16 +42,14 @@ object CommandUtils extends Logging {
 
   /** Change statistics after changing data by commands. */
   def updateTableStats(sparkSession: SparkSession, table: CatalogTable): Unit = {
-    if (table.stats.nonEmpty) {
```
Because of this condition check, the insert table command will never able to calculate the table size even if user enables sparkSession.sessionState.conf.autoSizeUpdateEnabled.

dongjoon-hyun replied:
Thank you for pinging me, @sujith71955 . Got it. I'll take a look. BTW, I update the JIRA ID in the title.

sujith71955 replied:
Sure . thanks

AmplabJenkins commented:
Can one of the admins verify this patch?

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/9694/
Test PASSed.

sujith71955 commented:
cc @dongjoon-hyun @HyukjinKwon @cloud-fan

sujith71955 commented:
Please review and let me know for any suggestions/clarifications. Thanks

SparkQA commented:
**[Test build #104362 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/104362/testReport)** for PR 24315 at commit [`b143e84`](https://github.com/apache/spark/commit/b143e845f9281c44d69900282b32e6ac99eb2cb0).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/104362/
Test FAILed.

dongjoon-hyun commented on sql/core/src/test/scala/org/apache/spark/sql/StatisticsCollectionSuite.scala:
```diff
@@ -337,6 +337,22 @@ class StatisticsCollectionSuite extends StatisticsCollectionTestBase with Shared
+    val autoUpdate = true
+      withSQLConf(SQLConf.AUTO_SIZE_UPDATE_ENABLED.key -> autoUpdate.toString) {
+        withTable(table) {
+          sql(s"CREATE TABLE $table (i int, j string) STORED AS PARQUET")
```
`STORE AS` is only supported in `hive` module. Please use `USING PARQUET`.

sujith71955 replied:
Right, I updated the same. thanks for the input.

wangyum reviewed (commented):
Please see the previous discussion: https://github.com/apache/spark/pull/20430/files#r164662154

sujith71955 committed:
[SPARK-274034][SQL] Table Statisics shall be updated automatically if auto update feature is enabled(spark.sql.statistics.size.autoUpdate.enabled =true)

What changes were proposed in this pull request?

For the  table, INSERT OVERWRITE command statistics are automatically computed by default if user set spark.sql.statistics.size.autoUpdate.enabled =true and the statistics shall be recorded
in metadata store, this is not happening currently  because of  validation table.stats.nonEmpty, the statistics were never recorded for the newly created table, this check  doesn't holds good if auto update property feature is enabled by the user.
As part of fix the autoSizeUpdateEnabled has been pulled up as part of separate validation which will ensure if this feature is enabled the system will calculate the size of the table in every insert command
and the same will be recorded in meta-store.

How was this patch tested?
UT is written and manually verified in cluster.

Tested with unit tests + some internal tests on real cluster.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/9704/
Test PASSed.

SparkQA commented:
**[Test build #104371 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/104371/testReport)** for PR 24315 at commit [`f87639e`](https://github.com/apache/spark/commit/f87639eb1775f88884d3c2180543bd48ba68e4dd).

sujith71955 commented:
> Please see the previous discussion: https://github.com/apache/spark/pull/20430/files#r164662154

I am not removing table.stats.nonEmpty validation here, it is intact.  The scenario mentioned by Zhenhua is a valid concern and this PR doesn't have any impact on the scenario what he mentioned.
  if user enables the auto size update feature, user expects the size will be calculated by the system automatically which was not happening , and this PR addressed the issue,  other logics are intact.  Thanks

sujith71955 commented:
> > Please see the previous discussion: https://github.com/apache/spark/pull/20430/files#r164662154
> 
> @wangyum : I am not removing table.stats.nonEmpty validation here, it is intact. The scenario mentioned by Zhenhua is a valid concern and this PR doesn't have any impact on the scenario what he mentioned.
> if user enables the auto size update feature, user expects the size will be calculated by the system automatically which was not happening , and this PR addressed the issue, other logics are intact. Thanks

Hope i clarified your point.

SparkQA commented:
**[Test build #104371 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/104371/testReport)** for PR 24315 at commit [`f87639e`](https://github.com/apache/spark/commit/f87639eb1775f88884d3c2180543bd48ba68e4dd).
 * This patch **fails due to an unknown error code, -9**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/104371/
Test FAILed.

sujith71955 commented:
retest this please

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/9730/
Test PASSed.

SparkQA commented:
**[Test build #104398 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/104398/testReport)** for PR 24315 at commit [`f87639e`](https://github.com/apache/spark/commit/f87639eb1775f88884d3c2180543bd48ba68e4dd).

SparkQA commented:
**[Test build #104398 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/104398/testReport)** for PR 24315 at commit [`f87639e`](https://github.com/apache/spark/commit/f87639eb1775f88884d3c2180543bd48ba68e4dd).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/104398/
Test FAILed.

sujith71955 commented:
> **[Test build #104398 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/104398/testReport)** for PR 24315 at commit [`f87639e`](https://github.com/apache/spark/commit/f87639eb1775f88884d3c2180543bd48ba68e4dd).
> 
> * This patch **fails PySpark unit tests**.
> * This patch merges cleanly.
> * This patch adds no public classes.

Seems to be a false alarm, Not relevant to this PR, i will re-trigger the build.

sujith71955 commented:
retest this please

SparkQA commented:
**[Test build #104419 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/104419/testReport)** for PR 24315 at commit [`f87639e`](https://github.com/apache/spark/commit/f87639eb1775f88884d3c2180543bd48ba68e4dd).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/9750/
Test PASSed.

SparkQA commented:
**[Test build #104419 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/104419/testReport)** for PR 24315 at commit [`f87639e`](https://github.com/apache/spark/commit/f87639eb1775f88884d3c2180543bd48ba68e4dd).
 * This patch **fails due to an unknown error code, -9**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/104419/
Test FAILed.

sujith71955 commented:
retest this please

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/9759/
Test PASSed.

SparkQA commented:
**[Test build #104428 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/104428/testReport)** for PR 24315 at commit [`f87639e`](https://github.com/apache/spark/commit/f87639eb1775f88884d3c2180543bd48ba68e4dd).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/104428/
Test PASSed.

sujith71955 commented:
@dongjoon-hyun  fixed the review comment  please let me know for any further clarifications/suggestion. thanks

dongjoon-hyun commented on sql/core/src/test/scala/org/apache/spark/sql/StatisticsCollectionSuite.scala:
```diff
@@ -337,6 +337,22 @@ class StatisticsCollectionSuite extends StatisticsCollectionTestBase with Shared
+  test("auto gather stats after insert command") {
+    val table = "change_stats_insert_datasource_table"
+    val autoUpdate = true
+      withSQLConf(SQLConf.AUTO_SIZE_UPDATE_ENABLED.key -> autoUpdate.toString) {
```
Thank you for updating, @sujith71955 . Here are two issues.
1. We need to test both cases by using `Seq(false, true).foreach { autoUpdate =>`
2. The current indentation is wrong at 343.

If you fix (1), (2) will be fixed together.

sujith71955 replied:
Handled the comment. thanks

dongjoon-hyun commented on sql/core/src/test/scala/org/apache/spark/sql/StatisticsCollectionSuite.scala:
```diff
@@ -337,6 +337,22 @@ class StatisticsCollectionSuite extends StatisticsCollectionTestBase with Shared
+          // analyze to get initial stats
+          // insert into command
+          sql(s"INSERT INTO TABLE $table SELECT 1, 'abc'")
+            val stats = getCatalogTable(table).stats
```
Also, fix the indentation here, too.

sujith71955 replied:
Fixed. thanks

sujith71955 committed:
[SPARK-274034][SQL] Table Statisics shall be updated automatically if auto update feature is enabled(spark.sql.statistics.size.autoUpdate.enabled =true)

What changes were proposed in this pull request?

For the  table, INSERT OVERWRITE command statistics are automatically computed by default if user set spark.sql.statistics.size.autoUpdate.enabled =true and the statistics shall be recorded
in metadata store, this is not happening currently  because of  validation table.stats.nonEmpty, the statistics were never recorded for the newly created table, this check  doesn't holds good if auto update property feature is enabled by the user.
As part of fix the autoSizeUpdateEnabled has been pulled up as part of separate validation which will ensure if this feature is enabled the system will calculate the size of the table in every insert command
and the same will be recorded in meta-store.

How was this patch tested?
UT is written and manually verified in cluster.

Tested with unit tests + some internal tests on real cluster.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/9819/
Test PASSed.

SparkQA commented:
**[Test build #104489 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/104489/testReport)** for PR 24315 at commit [`3692775`](https://github.com/apache/spark/commit/3692775ab281be2434e8e4af916e7a655f0d0377).

SparkQA commented:
**[Test build #104489 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/104489/testReport)** for PR 24315 at commit [`3692775`](https://github.com/apache/spark/commit/3692775ab281be2434e8e4af916e7a655f0d0377).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/104489/
Test PASSed.

sujith71955 commented:
@dongjoon-hyun All comments are handled. Thanks a lot for the valuable inputs. let me know for any further suggestions/clarifications

dongjoon-hyun reviewed (approved):
+1, LGTM. Merged to master. Thank you, @sujith71955 .

cc @gatorsmile and @cloud-fan

dongjoon-hyun commented:
Merged to branch-2.4, too.
